library/core/src/option.rs

```diff
@@ -70,10 +70,22 @@
 //! }
 //! ```
 //!
-//! This usage of [`Option`] to create safe nullable pointers is so
-//! common that Rust does special optimizations to make the
-//! representation of [`Option`]`<`[`Box<T>`]`>` a single pointer. Optional pointers
-//! in Rust are stored as efficiently as any other pointer type.
+//! # Representation
+//!
+//! Rust guarantees to optimize the following types `<T>` such that an
+//! [`Option<T>`] has the same size as `T`:
+//!
+//! * [`Box<T>`]
+//! * `&T`
+//! * `&mut T`
+//! * `extern "C" fn`
+//! * [`num::NonZero*`]
+//! * [`ptr::NonNull<T>`]
+//! * `#[repr(transparent)]` struct around one of the types in this list.
+//!
+//! For the above cases, it is guaranteed that one can [`mem::transmute`]
+//! from all valid values of `T` to `Option<T>` but only from non-`None`
```

Comment:
Why not use `Some::<T>` ?

Reply:
Agreed, positive statements are better than negative ones.

Reply:
Good question. I tend to agree that this is clearer, but also worried that people might not twig that `None` is bad. I've tried to cover both bases in https://github.com/rust-lang/rust/pull/75454/commits/f3d7196caec3f54e572c7389b1cef9fd9e62c1ed: comments welcome!